### DISCUSSION TITLE
명령어 사이클

### DISCUSSION THREAD

**Discussion by le2sky** [2024-woowacourse-study/level-interview#257] [Computer Architecture]
.

**Comment by le2sky** [comment]
CPU는 명령어를 실행하기 위해 다음과 같은 요소들이 필요하다.

- **프로그램 카운터 (Program Counter: PC)** : 다음에 인출될 명령어의 주소를 가지고 있다.  명령어가 인출된 후 자동으로 증가하거나, 분기의 경우 갱신된다.
- **누산기 (Accumulator: AC)** : 데이터를 일시적으로 저장한다. 레지스터 비트 수는 CPU가 한 번에 연산 처리할 수 있는 데이터 비트의 수(word)와 같다.
- **명령어 레지스터 (Instruction Register: IR)** : 가장 최근에 인출된 명령어가 저장되어 있다.
- **기억장치 주소 레지스터 (Memory Address Register: MAR)** : PC에 저장된 명령어 주소가 시스템 주소 버스로 출력되기 전에 일시적으로 저장된다. MAR 출력 선들이 주소 버스 선들과 직접 접속한다.
- **기억장치 버퍼 레지스터 (Memory Buffer Register: MBR)** : 기억장치에 저장될 데이터 혹은 기억장치로부터 읽혀진 데이터가 일시적으로 저장된다. MBR 입력 및 출력 선들이 데이터 버스 선들과 직접 접속 된다.
- **산술 논리 연산 장치 (Arithmetic and Logical Unit: ALU)** :  산술 및 논리 연산들을 수행하는 회로들로 이루어진 하드웨어 모듈이다.
- **제어 장치 (Control Unit: CU)** : 명령어를 해석하고, 실행하기 위한 제어 신호들을 순차적으로 발생한다. (명령어 실행에 필요한 정보들의 전송 통로 및 방향 지정, CPU 내부 요소들과 시스템 구성 요소들의 동작 시간도 결정함)

CPU는 기억장치에 저장되어 있는 프로그램 코드인 명령어들을 실햄함으로 프로그램 수행이라는 컴퓨터의 기본적인 기능을 수행한다. 이때, 한 명령어를 실행하는데 필요한 전체 과정을 명령어 사이클이라고 한다. 명령어 사이클은 다음 단계들로 이뤄진다. 

- **명령어 인출(instruction fetch)**
- **명령어 실행(instruction execution)**

이 단계들은 각각 부사이클(subsycle)로 구분하여 인출 사이클과 실행 사이클로 나눌 수 있다. 

**인출 사이클(fetch cycle)**

기억장치의 지정된 위치로부터 명령어를 읽어오는 과정을 인출 사이클이라고 한다.

```
T₀ : MAR <- PC
T₁ : MBR <- M[MAR], PC <- PC + 1
T₂ : IR <- MBR
```
T₀ : 현재 PC 내용을 CPU 내부 버스를 통하여 MAR로 보낸다. 시스템 주소 버스와 직접 접속된 MAR을 통해 주소가 기억장치로 전송된다.
T₁ : 주소가 지정하는 기억장치 위치로부터 읽혀진 명령어가 데이터 버스를 통해 MBR로 적재되며, PC의 내용을 업데이트에 다음 명령어 주소를 가리키게 한다.
T₂ : MBR에 저장되어 있는 명령어가 IR로 이동된다. 

**실행 사이클(execute cycle)**

CPU가 명령어를 해독하고 그 결과에 따라 필요한 연산들을 수행하는 과정을 실행 사이클이라고 한다.
이는 다음과 같이 네 가지로 분류된다.

- 데이터 이동 : CPU와 기억장치 혹은 CPU와 IO 장치 간에 데이터를 이동한다.
- 데이터 처리 : 데이터에 대하여 산술 혹은 논리 연산을 수행
- 데이터 저장 : 연산결과 데이터 혹은 입력장치로부터 읽어들인 데이터를 기억장치에 저장한다.
- 프로그램 제어 : 프로그램의 실행 순서를 결정한다.

데이터 이동의 예시 -> 기억 장치에 저장되어 있는 데이터를 이동(LOAD addr)
```
T₀ : MAR <- IR(addr)
T₁ : MBR <- M[MAR]
T₂ : AC <- MBR
```

데이터 저장의 예시 -> AC 내용을 기억장치에 저장 (STA addr)
```
T₀ : MAR <- IR(addr)
T₁ : MBR <- AC
T₂ : M[MAR] <- MBR
```

데이터 처리의 예시 -> 기억장치 저장 데이터를 AC와 더하고 다시 AC에 저장 (ADD addr)
```
T₀ : MAR <- IR(addr)
T₁ : MBR <- M[MAR]
T₂ : AC <- AC + MBR
```

프로그램 제어의 예시 -> JUMP addr
```
T₀ : PC <- IR(addr)
```

추가로 인터럽트 사이클과 간접 사이클이 존재한다.

**간접 사이클(indirect cycle)** 
- 간접 주정 방식은 오퍼랜드 필드에 유효 주소의 주소를 명시한다. 
- 이때, 명령어를 인출하여 CPU에 가져왔다고 해도 바로 실행 사이클에 돌입할 수 없다.
- 명령어를 실행하기 위해서는 메모리 접근을 한 번 더 해야 한다. 
- 이 단계를 간접 사이클이라고 한다.

**인터럽트 사이클(interupt cycle)**
- 입출력장치는 CPU 인터럽트 요청 신호를 보낸다.
- CPU는 실행 사이클이 끝나고 명령어를 인출하기 전 인터럽트 여부를 확인한다.
- CPU는 인터럽트 요청을 확인하고 인터럽트 플래그를 통해 현재 인터럽트 허용 여부를 확인한다.
- 인터럽트를 받아들일 수 있다면 CPU는 현재 작업을 백업한다.
- CPU는 인터럽트 벡터를 참조하여 인터럽트 서비스 루틴을 실행한다.
- 인터럽트 서비스 루틴 실행이 끝나면 백업해 둔 작업을 복구하여 실행을 재개 한다.
- 이러한 과정을 인터럽트 사이클이라고 한다.

**Comment by 3Juhwan** [comment]
CPU는 프로그램의 각 명령어를 일정 주기를 반복하며 실행한다. 이 주기를 명령어 사이클이라고 한다. 
명령어 사이클은 인출, 실행, 간접, 인터럽트 사이클로 구성되어 있다. 
<div align="center">
<img src="https://github.com/user-attachments/assets/6309f430-edaf-4c9a-a9ef-840707443642" alt="image" width="500">
</div>

</br> 

- 인출 사이클: 메모리에 적재된 명령어를 CPU로 가져오는 단계다. 
- 실행 사이클: 메모리에서 가져온 명령어를 CPU가 실행하는 단계다. 
- 간접 사이클: 인출 사이클에서 가져온 정보가 명령어가 아닌 간접 주소라면, 실제 주소를 가져오기 위해 메모리에 다시 접근해야 한다. 메모리에 다시 접근해서 유효 주소를 가져오는 단계다. 
- 인터럽트 사이클: 하드웨어 인터럽트가 발생하여 인터럽트 서비스 루틴을 실행하는 단계다. 인터럽트 서비스 루틴을 실행하기 전에 프로그램 카운터 값 등 현재 프로그램을 재개하기 위해 필요한 모든 내용을 백업해야 한다. 인터럽트 서비스 루틴이 실행된 후에 백업한 내용을 다시 불러와서 프로그램을 재개한다.
